lua/bufferline/options.lua

```diff
@@ -42,7 +51,82 @@ local DEFAULT_DIAGNOSTICS = {
+--- @field custom_color? boolean if present, this color will be used for ALL filetype icons
+--- @field enabled? boolean iff `true`, show the `devicons` for the associated buffer's `filetype`.
+
+--- @class bufferline.options.icons.buffer
+--- @field buffer_index? boolean iff `true`, show the index of the associated buffer with respect to the ordering of the buffers in the tabline.
+--- @field buffer_number? boolean iff `true`, show the `bufnr` for the associated buffer.
+--- @field button? string the button which is clicked to close / save a buffer, or indicate that it is pinned.
+--- @field filetype? bufferline.options.icons.filetype filetype icon options
+--- @field separator? string the left-hand separator between buffers in the tabline
+
+--- @class bufferline.options.icons.state: bufferline.options.icons.buffer
+--- @field modified? bufferline.options.icons.buffer the icons used for an modified buffer
+--- @field pinned? bufferline.options.icons.buffer the icons used for a pinned buffer
+
+--- @class bufferline.options.icons: bufferline.options.icons.state
+--- @field alternate? bufferline.options.icons.state the icons used for an alternate buffer
+--- @field current? bufferline.options.icons.state the icons for the current buffer
+--- @field inactive? bufferline.options.icons.state the icons for inactive buffers
+--- @field visible? bufferline.options.icons.state the icons for visible buffers
+
+--- @alias bufferline.options.icons.preset boolean|"both"|"buffer_number_with_icon"|"buffer_numbers"|"numbers"
+
+--- @type {[bufferline.options.icons.preset]: bufferline.options.icons}
+local PRESETS = {
+  [false] = {
+    buffer_number = false,
+    buffer_index = false,
+    filetype = {enabled = false},
+  },
+  [true] = {
+    buffer_number = false,
+    buffer_index = false,
+    filetype = {enabled = true},
+  },
+  both = {
+    buffer_index = true,
+    buffer_number = false,
+    filetype = {enabled = true},
+  },
+  buffer_number_with_icon = {
+    buffer_index = false,
+    buffer_number = true,
+    filetype = {enabled = true},
+  },
+  buffer_numbers = {
+    buffer_index = false,
+    buffer_number = true,
+    filetype = {enabled = false},
+  },
+  numbers = {
+    buffer_index = true,
+    buffer_number = false,
+    filetype = {enabled = false},
+  },
+}
+
+--- @type bufferline.options.icons
+local DEFAULT_ICONS = vim.tbl_extend('keep', PRESETS[true], {
+  button = '',
+  inactive = {separator = '▎'},
+  modified = {button = '●'},
+  pinned = {button = ''},
+  separator = '▎',
+})
+
+--- A table of options that used to exist, and where they are located now.
+--- @type {[string]: string[]}
+local DEPRECATED_ICON_OPTIONS = {
+  icon_close_tab = {'button'},
+  icon_close_tab_modified = {'modified', 'button'},
+  icon_custom_colors = {'filetype', 'custom_color'},
+  icon_pinned = {'pinned', 'button'},
+  icon_separator_active = {'separator'},
+  icon_separator_inactive = {'inactive', 'separator'},
```

Comment:
These options, along with `closable`, are obsolete with the introduction of this much more incrementally customizable `icons`. I have marked them as deprecated, but the options will _still work_ (i.e. they will be translated to the new schema).

Do you think we should mark them as deprecated, or should we just silently do the transformations? My only concern would be **_if_** there is ever a v2 in the future that these options wouldn't have had a deprecation period.